examples/ShowPlayer.cpp

```diff
@@ -66,90 +69,184 @@ int ShowPlayer::Init() {
+                  << m_playback_pos << " ms)";
+        return state;
+      case ShowLoader::INVALID_LINE:
+        HandleInvalidLine();
+        return state;
+      default: {
+      }
+    }
+    playhead_time += entry.next_wait;
+    entries[entry.universe] = entry;
+    if (!found && playhead_time == seek_time) {
+      // Use the next frame if landing on the trailing edge of a frame's timeout
+      found = true;
+      continue;
+    }
+    if (found || playhead_time > seek_time) {
+      break;
+    }
+  }
+  m_playback_pos = seek_time;
+
+  // Send data in the state it would be in at the given time
+  unordered_map<unsigned int, ShowEntry>::iterator entry_it;
+  for (entry_it = entries.begin(); entry_it != entries.end(); ++entry_it) {
+    SendFrame(entry_it->second);
   }
+  // Adjust the timeout to handle landing in the middle of the entry's timeout
+  RegisterNextTimeout(playhead_time-seek_time);
+
+  return ShowLoader::OK;
 }
 
 
 /**
- * Get the next time offset
+ * Send the next frame in the show file
  */
-ShowLoader::State ShowPlayer::RegisterNextTimeout() {
-  unsigned int timeout;
-  ShowLoader::State state = m_loader.NextTimeout(&timeout);
-  if (state != ShowLoader::OK) {
-    return state;
+void ShowPlayer::SendNextFrame() {
+  ShowEntry entry;
+  ShowLoader::State state = m_loader.NextEntry(&entry);
+  if (state == ShowLoader::END_OF_FILE ||
+      (m_stop > 0 && m_playback_pos >= m_stop)) {
+    if (entry.buffer.Size() > 0) {
+      // Send the last frame before looping/exiting
+      SendFrame(entry);
+    }
+    HandleEndOfFile();
+    return;
+  } else if (state == ShowLoader::INVALID_LINE) {
+    HandleInvalidLine();
+    return;
   }
+  SendEntry(entry);
+}
 
+
+/**
+ * Send @p entry, update playhead position, and wait for next
+ * @param entry the show file entry to send
+ */
+void ShowPlayer::SendEntry(const ShowEntry &entry) {
+  // Send DMX data
+  SendFrame(entry);
+
+  // Set when next to send data
+  m_playback_pos += entry.next_wait;
+  RegisterNextTimeout(entry.next_wait);
+}
+
+
+/**
+ * Send the next frame in @p timeout milliseconds
+ */
+void ShowPlayer::RegisterNextTimeout(const unsigned int timeout) {
   OLA_INFO << "Registering timeout for " << timeout << "ms";
   m_client.GetSelectServer()->RegisterSingleTimeout(
-      timeout,
-      ola::NewSingleCallback(this, &ShowPlayer::SendNextFrame));
-  return state;
+     timeout,
+     ola::NewSingleCallback(this, &ShowPlayer::SendNextFrame));
+}
+
+
+/**
+ * Send data contained in @p entry
+ */
+void ShowPlayer::SendFrame(const ShowEntry &entry) const {
+  OLA_INFO << "Universe: " << entry.universe << ": " << entry.buffer.ToString();
+  ola::client::SendDMXArgs args;
+  m_client.GetClient()->SendDMX(entry.universe, entry.buffer, args);
 }
 
 
 /**
  * Handle the case where we reach the end of file
  */
 void ShowPlayer::HandleEndOfFile() {
+  if (m_stop > m_playback_pos) {
+    OLA_WARN << "Show file ends before the stop time (Actual length "
```

Comment:
You could possibly delay until m_stop here to do what was asked for timing wise. I'm not sure if it's a good idea or can be handled with other options or flags.

Reply:
Hmmmm.  If I have a 10 second show file and want it to last 20 seconds, right now I can set delay to 10 and avoid the stop flag at all.  However, that also requires I do some math when computers are quite good at doing that for me.  It would also require first knowing the actual length of the show file using verify.

I feel like adding an extra option would sow more confusion, especially since the warning will always be seen with the default log level as warn.

Reply:
The more I think about the more I'm convinced waiting is the correct behavior.  Reusing the same scenario, keeping it as is could mean a show with start=1000 and end=5000 lasts for 2 seconds, for example.  With my end-user hat on, that feels like a bug.

Reply:
Yeah, I think we should log a warning during verify if you've passed in a stop flag which is bigger than the show file, but I think the reality will be you have a 9.9 second showfile and you want it to run for 10s and loop type of scenarios.